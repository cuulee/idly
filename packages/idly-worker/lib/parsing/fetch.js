var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as SphericalMercator from '@mapbox/sphericalmercator';
import { handleErrors } from 'helpers/promise';
import { stubXML } from 'parsing/fixtures';
const DOMParser = require('xmldom').DOMParser;
const merc = new SphericalMercator({
    size: 256
});
const fetchStub = () => {
    return Promise.resolve(stubXML);
};
const debug = false;
export function fetchTile(x, y, zoom) {
    return __awaiter(this, void 0, void 0, function* () {
        const xyz = [x, y, zoom].join(',');
        const bboxStr = merc.bbox(x, y, zoom).join(',');
        try {
            if (debug) {
                return new DOMParser().parseFromString(yield fetchStub(), 'text/xml');
            }
            let response = yield fetch(`https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`);
            response = handleErrors(response);
            const text = yield response.text();
            const parser = new DOMParser();
            const xml = parser.parseFromString(text, 'text/xml');
            return xml;
        }
        catch (e) {
            console.error(e);
            throw e;
        }
    });
}
//# sourceMappingURL=fetch.js.map