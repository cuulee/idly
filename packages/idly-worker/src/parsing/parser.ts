import { deepFreeze } from 'idly-common/lib/misc/deepFreeze';
import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { attributesGen } from 'idly-common/lib/osm/attributesGen';
import { genLngLat } from 'idly-common/lib/osm/genLngLat';
import { nodeFactory } from 'idly-common/lib/osm/nodeFactory';
import { relationFactory } from 'idly-common/lib/osm/relationFactory';
import { relationMemberGen } from 'idly-common/lib/osm/relationMemberGen';
import {
  Entity,
  EntityId,
  EntityTable,
  Node,
  ParentWays,
  Relation,
  Tags,
  Way
} from 'idly-common/lib/osm/structures';
import { tagsFactory } from 'idly-common/lib/osm/tagsFactory';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

export function calculateParentWays(parentWays: ParentWays, ways: Way[]) {
  console.time('Imm: calculateParentWays');
  const x = parentWays.withMutations(p => {
    ways.forEach(w => {
      w.nodes.forEach(nodeId => {
        p.update(nodeId, (s = ImSet()) => s.add(w.id));
      });
    });
  });
  console.timeEnd('Imm: calculateParentWays');
  return x;
}

const xmlIndex = process.env.NODE_ENV === 'test' ? 0 : 2;
/**
 * @NOTE parseXML doesnt care if there are duplicates
 *  as this helps in removing a particular tile
 *  and still maintaining the duplicated entity in other
 *  tile. If in case we started screening duplicates
 *  this might remove an entity which might span across
 *  multiple tiles.
 * @param xml
 * @param parentWays
 */
export function parseXML(
  xml: Document | undefined,
  parentWays: ParentWays = ImMap()
):
  | undefined
  | {
      entities: Entity[];
      parentWays: ParentWays;
    } {
  if (!xml || !xml.childNodes) return;
  const root = xml.childNodes[xmlIndex];

  const children = root.childNodes;
  const entities: Entity[] = [];
  const nodesXML: any = [];
  const waysXML: any = [];
  const relationsXML: any = [];
  const group: {
    node: Node[];
    way: Way[];
    relation: Relation[];
  } = {
    node: [],
    way: [],
    relation: []
  };
  for (let i = 0, l = children.length; i < l; i++) {
    const child = children[i];
    const parser = group[child.nodeName];
    if (parser) group[child.nodeName].push(child);
  }

  group.relation = group.relation.map(parsers.relation);
  group.way = group.way.map(w => parsers.way(w));

  // calculateParentWays right at this point allows us
  // to pass the parsed group.way which is needed to
  // calculate parentWays.
  // @NOTE calculateParentWays directly mutates
  //  I want to change this behaviour
  // parentWays = calculateParentWays(parentWays, group.way);

  parentWays = calculateParentWays(parentWays, group.way);

  group.node = group.node.map(n => parsers.node(n));

  return {
    entities: [...group.node, ...group.way, ...group.relation],
    parentWays
  };
}

function getVisible(attrs) {
  return (
    !attrs.getNamedItem('visible') ||
    attrs.getNamedItem('visible').value !== 'false'
  );
}

function getMembers(obj) {
  const elems = obj.getElementsByTagName('member');
  const members = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    members[i] = relationMemberGen({
      id: attrs.getNamedItem('type').value[0] + attrs.getNamedItem('ref').value,
      type: attrs.getNamedItem('type').value,
      role: attrs.getNamedItem('role').value
    });
  }
  return members;
}

function getNodes(obj) {
  const elems = obj.getElementsByTagName('nd');
  const nodes = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    nodes[i] = 'n' + elems[i].attributes.getNamedItem('ref').value;
  }
  return nodes;
}

function getTags(obj): Tags {
  const elems = obj.getElementsByTagName('tag');
  const t: Array<[string, string]> = [];
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    t.push([attrs.getNamedItem('k').value, attrs.getNamedItem('v').value]);
  }
  return tagsFactory(t);
}

function getLoc(attrs) {
  const lon = attrs.getNamedItem('lon') && attrs.getNamedItem('lon').value;
  const lat = attrs.getNamedItem('lat') && attrs.getNamedItem('lat').value;
  return genLngLat([parseFloat(lon), parseFloat(lat)]);
}

const parsers = {
  node: function nodeData(obj) {
    const attrs = obj.attributes;
    const id = 'n' + attrs.getNamedItem('id').value;
    return nodeFactory({
      id,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value
      }),
      loc: getLoc(attrs),
      tags: getTags(obj)
    });
  },

  way: function wayData(obj) {
    const attrs = obj.attributes;
    return wayFactory({
      id: 'w' + attrs.getNamedItem('id').value,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value
      }),
      tags: getTags(obj),
      nodes: getNodes(obj)
    });
  },

  relation: function relationData(obj) {
    const attrs = obj.attributes;
    return relationFactory({
      id: 'r' + attrs.getNamedItem('id').value,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value
      }),
      tags: getTags(obj),
      members: getMembers(obj)
    });
  }
};
