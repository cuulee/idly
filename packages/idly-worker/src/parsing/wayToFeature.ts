import * as turfLineString from 'turf-linestring';
import * as turfPolygon from 'turf-polygon';

import { EntityTable, Node, OsmGeometry, Way } from 'idly-common/lib';

import { LineString, Polygon } from 'geojson';

/**
 * @REVISIT this func also needs to handle modifiedGraph
 *  for now i removed it so i can test caching and anyway
 *  not yet implemented modified ways
 */

export function wayCombiner(
  way: Way,
  table: EntityTable,
  existingProps: { geometry: OsmGeometry; [key: string]: string } = {
    geometry: OsmGeometry.LINE
  }
) {
  // @TOFIX code duplication
  const geometry = existingProps.geometry;
  if (!geometry) {
    throw new Error('geometry not found in existing props');
  }
  const nodes = getCoordsFromTable(table, way.nodes);
  return {
    id: way.id,
    ...wayToLineString(geometry, nodes),
    properties: {
      ...existingProps,
      id: way.id
    }
  };
}

export function getCoordsFromTable(
  table: EntityTable,
  nodes: string[]
): number[][] {
  return nodes.map(n => {
    if (!table.has(n)) throw new Error('node not found ' + n);
    const node = table.get(n) as Node;
    return [node.loc.lon, node.loc.lat];
  });
}

/**
 * graphNode comes from graph.node
 *  planning to use it for memoization.
 *  every ms counts. hehehe.
 */
export function wayToLineString(geometry: OsmGeometry, nodeCoords: number[][]) {
  if (geometry === OsmGeometry.LINE) {
    return turfLineString(nodeCoords, {});
  } else if (geometry === OsmGeometry.AREA) {
    return turfPolygon([nodeCoords], {});
  } else {
    throw new Error('not a matching geometry provided for way');
  }
}
