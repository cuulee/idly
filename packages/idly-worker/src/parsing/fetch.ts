import * as SphericalMercator from '@mapbox/sphericalmercator';
import { handleErrors } from 'helpers/promise';
import { stubXML } from 'parsing/fixtures';
const DOMParser = require('xmldom').DOMParser;

const merc = new SphericalMercator({
  size: 256
});

const fetchStub = () => {
  return Promise.resolve(stubXML);
};
const debug = false;

export async function fetchTile(x: number, y: number, zoom: number) {
  const xyz = [x, y, zoom].join(',');
  const bboxStr = merc.bbox(x, y, zoom).join(',');
  try {
    if (debug) {
      return new DOMParser().parseFromString(await fetchStub(), 'text/xml');
    }
    let response = await fetch(
      `https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`
    );
    response = handleErrors(response);
    const text = await response.text();

    const parser = new DOMParser();
    const xml = parser.parseFromString(text, 'text/xml');
    return xml;
  } catch (e) {
    console.error(e);
    throw e;
  }
}
