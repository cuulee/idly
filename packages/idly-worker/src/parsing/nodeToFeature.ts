import { Feature, Point } from 'geojson';

import { getNodeGeometry } from 'helpers/getNodeGeometry';
import { ParentWays } from 'parsing/parser';
import { presetsMatcherCached } from 'presets/presets';
import { Geometry } from 'structs/geometry';
import { Node } from 'structs/node';
import { Tags } from 'structs/tags';

type NodeFeat = Feature<Point>;
type NodeGeometry = Geometry.POINT | Geometry.VERTEX | Geometry.VERTEX_SHARED;

interface INodeMarkup {
  icon: string;
  name?: string;
  geometry: NodeGeometry;
}
export const DEFAULT_NODE_ICON = 'circle';

export function nodeCombiner(node: Node, parentWays: ParentWays) {
  return {
    ...nodeToPoint(node),
    properties: {
      ...applyNodeMarkup(
        // presetsMatcher,
        getNodeGeometry(node.id, parentWays),
        node.tags
      ),
      id: node.id
    }
  };
}

export const nodeToPoint = (node: Node): Feature<Point> => {
  return {
    id: node.id,
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [node.loc.lon, node.loc.lat]
    },
    properties: {}
  };
};

export const applyNodeMarkup = (geometry: NodeGeometry, tags: Tags) => {
  const match = presetsMatcherCached(geometry)(tags);
  return {
    icon: (match && match.icon) || DEFAULT_NODE_ICON,
    name: tags.get('name'),
    geometry
  };
};
