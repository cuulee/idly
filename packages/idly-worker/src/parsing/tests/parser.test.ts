import { ParentWays } from 'idly-common/lib/osm/structures';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

import {
  miniWayXML1,
  miniXML1,
  miniXML2,
  miniXML3,
  nodeXML1
} from '../../parsing/fixtures';
import { calculateParentWays, parseXML } from '../../parsing/parser';

const wayXML =
  '<?xml version="1.0" encoding="UTF-8"?><osm>' +
  '<way id="1" visible="true" timestamp="2008-01-03T05:24:43Z" version="1" changeset="522559"><nd ref="1"/></way>' +
  '</osm>';

describe('parsers', () => {
  describe('node', () => {
    it('matches snapshot', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniXML2, 'text/xml');
      expect(parseXML(xml)).toMatchSnapshot();
      expect(
        parseXML(parser.parseFromString(miniXML1, 'text/xml'))
      ).toMatchSnapshot();
    });
    it('matches a bigger example', () => {
      const parser = new DOMParser();
      const XML = nodeXML1;
      const xml = parser.parseFromString(XML, 'text/xml');
      expect(parseXML(xml)).toMatchSnapshot();
    });
  });
  describe('way', () => {
    it('matches snapshot', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniWayXML1, 'text/xml');
      expect(parseXML(xml)).toMatchSnapshot();
    });
  });
  describe('calculateParentWays ', () => {
    const obj = {
      n3780767744: new Set(['w40542208']),
      n4558992269: new Set(['w40542208']),
      n253179996: new Set([
        'w40882200',
        'w237684574',
        'w173431854',
        'w450548831'
      ]),
      n1485636774: new Set(['w40882200', 'w135262258'])
    };

    const parentWays: ParentWays = new Map();
    Object.keys(obj).forEach(k => parentWays.set(k, obj[k]));

    it('works', () => {
      expect(calculateParentWays(new Map(), [])).toEqual(new Map());
    });
    it('takes existing parentways and returns them', () => {
      expect(calculateParentWays(parentWays, [])).toEqual(parentWays);
    });
    it('takes existing parentways and adds a way', () => {
      const w1 = wayFactory({ id: 'w', nodes: ['n'] });
      expect(calculateParentWays(parentWays, [w1]).get('n')).toEqual(
        new Set([w1.id])
      );
    });
    it('takes existing parentways and appends a way', () => {
      const w1 = wayFactory({ id: 'w', nodes: ['n253179996'] });
      expect(calculateParentWays(parentWays, [w1])).toMatchSnapshot();
    });
    it('takes handles words', () => {
      const w1 = wayFactory({ id: 'way-great', nodes: ['n-random'] });
      expect(calculateParentWays(parentWays, [w1]).get('n-random')).toEqual(
        new Set([w1.id])
      );
    });
  });
  describe('it makes good parentways', () => {
    it('matches snapshot', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniXML3, 'text/xml');
      expect(parseXML(xml)).toMatchSnapshot();
    });
    it('reuses parent Ways and adds stuff to it', () => {
      const parser = new DOMParser();
      const xml = parser.parseFromString(miniXML3, 'text/xml');
      expect(parseXML(xml)).toMatchSnapshot();
    });
  });
});
