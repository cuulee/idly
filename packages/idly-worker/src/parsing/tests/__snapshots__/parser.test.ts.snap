// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsers calculateParentWays  takes existing parentways and appends a way 1`] = `
Map {
  "n3780767744" => Set {
    "w40542208",
  },
  "n4558992269" => Set {
    "w40542208",
  },
  "n253179996" => Set {
    "w40882200",
    "w237684574",
    "w173431854",
    "w450548831",
    "w",
  },
  "n1485636774" => Set {
    "w40882200",
    "w135262258",
  },
  "n" => Set {
    "w",
  },
}
`;

exports[`parsers it makes good parentways matches snapshot 1`] = `
Object {
  "entities": Array [
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n1",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n2",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n3",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Map {},
      "type": "way",
    },
    Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w2",
      "nodes": Array [
        "n1",
        "n3",
      ],
      "tags": Map {},
      "type": "way",
    },
  ],
  "parentWays": Map {
    "n1" => Set {
      "w1",
      "w2",
    },
    "n2" => Set {
      "w1",
    },
    "n3" => Set {
      "w2",
    },
  },
}
`;

exports[`parsers it makes good parentways reuses parent Ways and adds stuff to it 1`] = `
Object {
  "entities": Array [
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n1",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n2",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n3",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Map {},
      "type": "way",
    },
    Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w2",
      "nodes": Array [
        "n1",
        "n3",
      ],
      "tags": Map {},
      "type": "way",
    },
  ],
  "parentWays": Map {
    "n1" => Set {
      "w1",
      "w2",
    },
    "n2" => Set {
      "w1",
    },
    "n3" => Set {
      "w2",
    },
  },
}
`;

exports[`parsers node matches a bigger example 1`] = `
Object {
  "entities": Array [
    Object {
      "attributes": Object {
        "changeset": "138701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n12",
      "loc": Object {
        "lat": 34.5280866,
        "lon": 74.1795357,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782712",
      "loc": Object {
        "lat": 34.5280866,
        "lon": 74.1795357,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782713",
      "loc": Object {
        "lat": 34.528206,
        "lon": 74.1807924,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782714",
      "loc": Object {
        "lat": 34.5283653,
        "lon": 74.1822907,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782715",
      "loc": Object {
        "lat": 34.528425,
        "lon": 74.1815174,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782716",
      "loc": Object {
        "lat": 34.5285445,
        "lon": 74.1788832,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782717",
      "loc": Object {
        "lat": 34.5290024,
        "lon": 74.1788349,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782718",
      "loc": Object {
        "lat": 34.5292015,
        "lon": 74.179584,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782719",
      "loc": Object {
        "lat": 34.529301,
        "lon": 74.1824357,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782720",
      "loc": Object {
        "lat": 34.5294603,
        "lon": 74.1794149,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782721",
      "loc": Object {
        "lat": 34.5296992,
        "lon": 74.1823873,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782722",
      "loc": Object {
        "lat": 34.5297191,
        "lon": 74.1788349,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782723",
      "loc": Object {
        "lat": 34.5300377,
        "lon": 74.1821698,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782724",
      "loc": Object {
        "lat": 34.5303562,
        "lon": 74.1783274,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782725",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782726",
      "loc": Object {
        "lat": 34.5305354,
        "lon": 74.1805024,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782727",
      "loc": Object {
        "lat": 34.5309137,
        "lon": 74.1799224,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782728",
      "loc": Object {
        "lat": 34.5311526,
        "lon": 74.1786899,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782729",
      "loc": Object {
        "lat": 34.5314114,
        "lon": 74.1793182,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w411770875",
      "nodes": Array [
        "n4133782719",
        "n4133782714",
        "n4133782715",
        "n4133782713",
        "n4133782712",
        "n4133782716",
        "n4133782717",
        "n4133782718",
        "n4133782720",
        "n4133782722",
        "n4133782724",
        "n4133782728",
        "n4133782729",
        "n4133782727",
        "n4133782726",
        "n4133782725",
        "n4133782723",
        "n4133782721",
        "n4133782719",
      ],
      "tags": Map {
        "landuse" => "residential",
        "place" => "village",
        "source" => "MSF Survey",
      },
      "type": "way",
    },
  ],
  "parentWays": Map {
    "n4133782719" => Set {
      "w411770875",
    },
    "n4133782714" => Set {
      "w411770875",
    },
    "n4133782715" => Set {
      "w411770875",
    },
    "n4133782713" => Set {
      "w411770875",
    },
    "n4133782712" => Set {
      "w411770875",
    },
    "n4133782716" => Set {
      "w411770875",
    },
    "n4133782717" => Set {
      "w411770875",
    },
    "n4133782718" => Set {
      "w411770875",
    },
    "n4133782720" => Set {
      "w411770875",
    },
    "n4133782722" => Set {
      "w411770875",
    },
    "n4133782724" => Set {
      "w411770875",
    },
    "n4133782728" => Set {
      "w411770875",
    },
    "n4133782729" => Set {
      "w411770875",
    },
    "n4133782727" => Set {
      "w411770875",
    },
    "n4133782726" => Set {
      "w411770875",
    },
    "n4133782725" => Set {
      "w411770875",
    },
    "n4133782723" => Set {
      "w411770875",
    },
    "n4133782721" => Set {
      "w411770875",
    },
  },
}
`;

exports[`parsers node matches snapshot 1`] = `
Object {
  "entities": Array [
    Object {
      "attributes": Object {
        "changeset": "1",
        "timestamp": "2009-03-07T03:26:33Z",
        "uid": null,
        "user": null,
        "version": "1",
        "visible": true,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Map {},
      "type": "node",
    },
  ],
  "parentWays": Map {},
}
`;

exports[`parsers node matches snapshot 2`] = `
Object {
  "entities": Array [
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782712",
      "loc": Object {
        "lat": 34.5280866,
        "lon": 74.1795357,
      },
      "tags": Map {},
      "type": "node",
    },
    Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:39Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n4133782713",
      "loc": Object {
        "lat": 34.528206,
        "lon": 74.1807924,
      },
      "tags": Map {},
      "type": "node",
    },
  ],
  "parentWays": Map {},
}
`;

exports[`parsers way matches snapshot 1`] = `
Object {
  "entities": Array [
    Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w411770875",
      "nodes": Array [
        "n4133782719",
        "n4133782714",
        "n4133782715",
        "n4133782713",
        "n4133782712",
        "n4133782716",
        "n4133782717",
        "n4133782718",
        "n4133782720",
        "n4133782722",
        "n4133782724",
        "n4133782728",
        "n4133782729",
        "n4133782727",
        "n4133782726",
        "n4133782725",
        "n4133782723",
        "n4133782721",
        "n4133782719",
      ],
      "tags": Map {
        "landuse" => "residential",
        "place" => "village",
        "source" => "MSF Survey",
      },
      "type": "way",
    },
  ],
  "parentWays": Map {
    "n4133782719" => Set {
      "w411770875",
    },
    "n4133782714" => Set {
      "w411770875",
    },
    "n4133782715" => Set {
      "w411770875",
    },
    "n4133782713" => Set {
      "w411770875",
    },
    "n4133782712" => Set {
      "w411770875",
    },
    "n4133782716" => Set {
      "w411770875",
    },
    "n4133782717" => Set {
      "w411770875",
    },
    "n4133782718" => Set {
      "w411770875",
    },
    "n4133782720" => Set {
      "w411770875",
    },
    "n4133782722" => Set {
      "w411770875",
    },
    "n4133782724" => Set {
      "w411770875",
    },
    "n4133782728" => Set {
      "w411770875",
    },
    "n4133782729" => Set {
      "w411770875",
    },
    "n4133782727" => Set {
      "w411770875",
    },
    "n4133782726" => Set {
      "w411770875",
    },
    "n4133782725" => Set {
      "w411770875",
    },
    "n4133782723" => Set {
      "w411770875",
    },
    "n4133782721" => Set {
      "w411770875",
    },
  },
}
`;
